store/src/java/com/zimbra/cs/event/logger/EventLogger.java

```diff
@@ -1,25 +1,299 @@
+    public static EventLogger getEventLogger(ConfigProvider config) {
+        synchronized (EventLogger.class) {
+            if (instance == null) {
+                instance = new EventLogger(config);
+            } else {
+                instance.setConfigProvider(config);
+            }
+        }
+        return instance;
+    }
+
+    private void setConfigProvider(ConfigProvider config) {
+        this.config = config;
+        restartEventNotifierExecutor();
+    }
+
+    public static void registerHandlerFactory(String factoryName, EventLogHandler.Factory factory) {
+        if(factoryMap.containsKey(factoryName)) {
+            ZimbraLog.event.warn("EventLogHandler Factory %s already registered", factoryName);
+        } else {
+            factoryMap.put(factoryName, factory);
+        }
+    }
+
+    public static boolean unregisterHandlerFactory(String factoryName) {
+        if(!factoryMap.containsKey(factoryName)) {
+            ZimbraLog.event.warn("EventLogHandler Factory %s is not registered", factoryName);
+            return false;
+        } else {
+            factoryMap.remove(factoryName);
+            return true;
+        }
+    }
+
+    @VisibleForTesting
+    public static void unregisterAllHandlerFactories() {
+        factoryMap.clear();
+    }
+
+    public static boolean isFactoryRegistered(String factoryName) {
+        return factoryMap.containsKey(factoryName);
+    }
+
+    /**
+     * Restart the executor service, picking up new configuration data
+     */
+    public void restartEventNotifierExecutor() {
+        shutdownEventNotifierExecutor();
+        startupEventNotifierExecutor();
+    }
 
-    protected static EventLogger eventLogger;
+    public void log(List<Event> events) {
+        for (Event event : events) {
+            log(event);
+        }
+    }
 
-    public static final void setEventLogger(Class<? extends EventLogger> loggerClass) throws ServiceException {
-        String className = loggerClass.getName();
-        ZimbraLog.event.info("setting EventLogger class %s", className);
+    public boolean log(Event event) {
         try {
-            eventLogger = loggerClass.newInstance();
-        } catch (InstantiationException | IllegalAccessException e) {
-            throw ServiceException.FAILURE(String.format("unable to initialize EventLogger setEventLogger %s", className), e);
+            return eventQueue.add(event);
+        } catch (IllegalStateException e) {
+            ZimbraLog.event.info("unable to add item for account %s to indexing queue", event.getAccountId());
+            return false;
         }
     }
 
-    public static EventLogger getEventLogger() {
-        return eventLogger;
+    private Map<String, Collection<String>> getConfigMap() {
+        Map<String, Collection<String>> configMap = config.getHandlerConfig();
+        configMap.keySet().retainAll(factoryMap.keySet());
+        return configMap;
+    }
+
+    public void startupEventNotifierExecutor() {
```

Comment:
Stylistic question:  Why is this section spaced out so much more than the rest?

Reply:
Varun and I collaborated on this file and I guess we have different spacing habits :)

Reply:
It might make sense to update http://wiki.eng.zimbra.com/index.php/Zimbra_Server_Code_Style and/or refer to that to keep code style consistency (more nits noted below).  Consistency is good.

Reply:
Done, reduced the spacing.